File: test_deep_render/js/test_deep_render.js
var Nested = React.createClass({
  render: function() {
    var level = this.props.level;
    if (level > 1) {
      return React.createElement(Nested, {level: level - 1});
    }

    return React.createElement('div', {});
  }
});

var testTypes = {
  'DEEP': 'DEEP',
  'SHALLOW': 'SHALLOW'
};

var TEST_TYPE;
var LEVELS;
var COUNT;
function runTest() {
  var i, t0, t1;
  var time, message;

  switch (TEST_TYPE) {
    case testTypes.DEEP:
      var mountNode = document.createElement('div');
      React.render(React.createElement(Nested, {level: LEVELS}), mountNode);
      t0 = performance.now();
      for (i=0; i<COUNT; i++) {
        React.render(React.createElement(Nested, {level: LEVELS}), mountNode);
      }
      t1 = performance.now();

      message = ["Rerendering ", LEVELS, " levels of components (sample size = ", COUNT, ") took "];
      break;

    case testTypes.SHALLOW:
      t0 = performance.now();
      for (i=0; i<COUNT; i++) {
        React.render(React.createElement(Nested, {level: LEVELS}), document.createElement('div'));
      }
      t1 = performance.now();

      message = ["Rendering ", LEVELS, " levels of components (sample size = ", COUNT, ") took "];
      break;
  }

  time = (t1 - t0) / COUNT;

  var status = document.createElement('div');
  React.render(React.DOM.div({},
    message,
    React.DOM.input({
      'style': {'fontFamily': 'monospace'},
      'readOnly': true,
      'value': time
    }),
    " milliseconds."
  ), status);
  document.body.appendChild(status);
}


var getQueryString = function ( field, url ) {
    var href = url ? url : window.location.href;
    var reg = new RegExp( '[?&]' + field + '=([^&#]*)', 'i' );
    var string = reg.exec(href);
    return string ? string[1] : null;
};

LEVELS = parseInt(getQueryString('levels'));
COUNT = parseInt(getQueryString('count'));
TEST_TYPE = (getQueryString('test') || '').toUpperCase();

if (isNaN(LEVELS) || isNaN(COUNT) || !(TEST_TYPE in testTypes)) {
  throw 'Invalid query params. Requires `levels`, `count`, and `test`.';
}

runTest();

File: test_deep_render/dart_cleandart_new_js/web/test_deep_render.dart
library test_deep_render;

import 'dart:html';
import 'package:react/react.dart' as react;
import 'package:react/react_client.dart';

ReactComponentFactory Nested = react.registerComponent(() => new NestedComponent());
class NestedComponent extends react.Component {
  render() {
    int level = props['level'];
    if (level > 1) {
      return Nested({'level': level - 1});
    }

    return react.div({});
  }
}

enum TestType {
  SHALLOW, DEEP
}

int COUNT;
int LEVELS;
TestType TEST_TYPE;
runTest() {
  double time;
  String message;

  switch(TEST_TYPE) {
    case TestType.SHALLOW:
      double t0 = window.performance.now();
      for (int i=0; i<COUNT; i++) {
        react.render(Nested({'level': LEVELS}), new DivElement());
      }
      double t1 = window.performance.now();

      time = (t1 - t0) / COUNT;
      message = "Initial render of $LEVELS levels of components (sample size = $COUNT) took ";
      break;

    case TestType.DEEP:
      // Initial render
      var div = new DivElement();
      react.render(Nested({'level': LEVELS}), div);

      // Rerender
      double t0 = window.performance.now();
      for (int i=0; i<COUNT; i++) {
        react.render(Nested({'level': LEVELS}), div);
      }
      double t1 = window.performance.now();

      time = (t1 - t0) / COUNT;
      message = "Rerendering $LEVELS levels of components (sample size = $COUNT) took ";
      break;
  }

  var status = document.createElement('div');
  react.render(react.div({},
    message,
    react.input({
      'style': {'fontFamily': 'monospace'},
      'readOnly': true,
      'value': time,
    }),
    " milliseconds."
  ), status);
  document.body.append(status);
}

main() {
  var params = Uri.parse(window.location.toString()).queryParameters;
  try {
    COUNT = int.parse(params['count']);
    LEVELS = int.parse(params['levels']);

    const Map testParamValueToTestType = const {
      'SHALLOW': TestType.SHALLOW,
      'DEEP': TestType.DEEP,
    };
    TEST_TYPE = testParamValueToTestType[params['test'].toUpperCase()];

    if (COUNT.isNaN || LEVELS.isNaN) {
      throw 'Invalid number';
    }
    if (TEST_TYPE == null) {
      throw 'Invalid test type';
    }
  } catch(e) {
    throw 'Invalid query params. Requires `levels`, `count`, and `test`.';
  }

  setClientConfiguration();
  runTest();
}
